File: koladata/operators/uuobj.cc
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include "koladata/operators/uuobj.h"

#include <memory>
#include <string>
#include <utility>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_format.h"
#include "absl/types/span.h"
#include "koladata/adoption_utils.h"
#include "koladata/data_bag.h"
#include "koladata/data_slice.h"
#include "koladata/data_slice_qtype.h"
#include "koladata/object_factories.h"
#include "koladata/operators/utils.h"
#include "arolla/memory/frame.h"
#include "arolla/qexpr/bound_operators.h"
#include "arolla/qexpr/eval_context.h"
#include "arolla/qexpr/operators.h"
#include "arolla/qexpr/qexpr_operator_signature.h"
#include "arolla/qtype/optional_qtype.h"
#include "arolla/qtype/qtype.h"
#include "arolla/qtype/qtype_traits.h"
#include "arolla/qtype/typed_slot.h"
#include "arolla/util/repr.h"
#include "arolla/util/text.h"
#include "arolla/util/status_macros_backport.h"

namespace koladata::ops {
namespace {

class UuObjOperator : public arolla::QExprOperator {
 public:
  explicit UuObjOperator(absl::Span<const arolla::QTypePtr> input_types)
      : QExprOperator("kde.core._uuobj",
                      arolla::QExprOperatorSignature::Get(
                          input_types, arolla::GetQType<DataSlice>())) {}

  absl::StatusOr<std::unique_ptr<arolla::BoundOperator>> DoBind(
      absl::Span<const arolla::TypedSlot> input_slots,
      arolla::TypedSlot output_slot) const final {
    return arolla::MakeBoundOperator(
        [seed_slot = input_slots[0].UnsafeToSlot<DataSlice>(),
         named_tuple_slot = input_slots[1],
         output_slot = output_slot.UnsafeToSlot<DataSlice>()](
            arolla::EvaluationContext* ctx, arolla::FramePtr frame) {
          const auto& seed_data_slice = frame.Get(seed_slot);
          if (seed_data_slice.GetShape().rank() != 0 ||
              !seed_data_slice.item().holds_value<arolla::Text>()) {
            ctx->set_status(absl::InvalidArgumentError(absl::StrFormat(
                "requires seed to be DataItem holding Text, got %s",
                arolla::Repr(seed_data_slice))));
            return;
          }
          auto seed = seed_data_slice.item().value<arolla::Text>();
          auto attr_names = GetAttrNames(named_tuple_slot);
          auto values = GetValueDataSlices(named_tuple_slot, attr_names, frame);
          auto db = koladata::DataBag::Empty();
          koladata::AdoptionQueue adoption_queue;
          for (const auto &ds : values) {
            adoption_queue.Add(ds);
          }
          auto status = adoption_queue.AdoptInto(*db);
          if (!status.ok()) {
            ctx->set_status(std::move(status));
            return;
          }
          ASSIGN_OR_RETURN(auto result,
                           UuObjectCreator()(db, seed, attr_names, values),
                           ctx->set_status(std::move(_)));
          frame.Set(output_slot, std::move(result));
        });
  }
};

}  // namespace

absl::StatusOr<arolla::OperatorPtr> UuObjOperatorFamily::DoGetOperator(
    absl::Span<const arolla::QTypePtr> input_types,
    arolla::QTypePtr output_type) const {
  if (input_types.size() != 2) {
    return absl::InvalidArgumentError("requires exactly 2 arguments");
  }
  if (input_types[0] != arolla::GetQType<DataSlice>()) {
    return absl::InvalidArgumentError(
        "requires first argument to be DataSlice");
  }
  RETURN_IF_ERROR(VerifyNamedTuple(input_types[1]));
  return arolla::EnsureOutputQTypeMatches(
      std::make_shared<UuObjOperator>(input_types), input_types, output_type);
}

}  // namespace koladata::ops

File: py/koladata/operators/uuobj.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""UUObj operators."""

from arolla import arolla
from koladata.operators import optools
from koladata.operators import qtype_utils
from koladata.types import data_slice
from koladata.types import qtypes

M = arolla.M
P = arolla.P
constraints = arolla.optools.constraints


# TODO: Make this operator public again once non-determinism is
# designed and implemented.
@optools.add_to_registry()
@optools.as_backend_operator(
    'kde.core._uuobj',
    aux_policy='koladata_obj_kwargs',
    qtype_constraints=[
        qtype_utils.expect_data_slice(P.seed),
        (
            M.qtype.is_namedtuple_qtype(P.kwargs),
            f'expected named tuple, got {constraints.name_type_msg(P.kwargs)}',
        ),
        qtype_utils.expect_data_slice_kwargs(P.kwargs),
    ],
    qtype_inference_expr=qtypes.DATA_SLICE
)
def _uuobj(seed=data_slice.DataSlice.from_vals(''), kwargs=arolla.namedtuple()):  # pylint: disable=unused-argument
  """Creates Object(s) whose ids are uuid(s) with the provided attributes.

  In order to create a different id from the same arguments, use
  `seed` argument with the desired value, e.g.

  kd.uuobj(seed='type_1', x=[1, 2, 3], y=[4, 5, 6])

  and

  kd.uuobj(seed='type_2', x=[1, 2, 3], y=[4, 5, 6])

  have different ids.

  Args:
    seed: text seed for the uuid computation.
    kwargs: a named tuple mapping attribute names to DataSlices. The DataSlice
      values must be alignable.

  Returns:
    (DataSlice) of uuids. The provided attributes are also set in a newly
    created databag.
  """
  raise NotImplementedError('implemented in the backend')
